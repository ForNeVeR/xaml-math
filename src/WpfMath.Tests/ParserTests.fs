module WpfMath.Tests.ParserTests

open Xunit

open WpfMath.Tests.ApprovalTestUtils

[<Fact>]
let ``2+2``() =
    verifyParseResult "2+2"

[<Theory>]
[<InlineData("(", ")", "(", ")")>]
[<InlineData("[", "]", "lbrack", "rbrack")>]
[<InlineData("{", "}", "lbrace", "rbrace")>]
[<InlineData("<", ">", "langle", "rangle")>]
let delimiters(left : string, right : string, lResult : string, rResult : string) =
    verifyParseResultScenario
    <| sprintf "%s,%s" lResult rResult
    <| sprintf @"\left%sa\right%s" left right

[<Theory>]
[<InlineData(".", ")", true, false)>]
[<InlineData("(", ".", false, true)>]
let emptyDelimiters(left : string, right : string, isLeftEmpty : bool, isRightEmpty : bool) =
    verifyParseResultScenario
    <| sprintf "(%s,%s,%A,%A)" left right isLeftEmpty isRightEmpty
    <| sprintf @"\left%sa\right%s" left right

[<Fact>]
let unmatchedDelimiters() =
    verifyParseResult @"\left)a\right|"

[<Fact>]
let expressionInBraces() =
    verifyParseResult @"\left(2+2\right)"

[<Fact>]
let expressionAfterBraces() =
    verifyParseResult @"\left(2+2\right) + 1"

[<Fact>]
let textCommand() =
    verifyParseResult @"\text{test}"

[<Fact>]
let spacesInText() =
    verifyParseResult @"\text{a b c}"

[<Fact>]
let cyrillicText() =
    verifyParseResult @"\text{абв}"

[<Fact>]
let underscoreText() =
    verifyParseResult @"\text{_}"

[<Fact>]
let commandsInText() =
    verifyParseResult @"\text{\alpha \beta \unknowncommand}"

[<Fact>]
let mathrm() =
    verifyParseResult @"\mathrm{sin}"

[<Fact>]
let complexMathrm() =
    verifyParseResult @"\mathrm{\left(2+2\right)} + 1"

[<Fact>]
let mathit() =
    verifyParseResult @"\mathit{sin}"

[<Fact>]
let mathcal() =
    verifyParseResult @"\mathcal{sin}"

[<Fact>]
let lim() =
    verifyParseResult @"\lim_{n} x"

[<Fact>]
let limInCurlyBraces() =
    verifyParseResult @"{\lim} x"

[<Fact>]
let sin() =
    verifyParseResult @"\sin^{n} x"

[<Fact>]
let intF() =
    verifyParseResult @"\int f"

[<Fact>]
let emptyCurlyBraces() =
    verifyParseResult @"{}"

[<Fact>]
let delimiterWithScripts() =
    verifyParseResult @"\left(2+2\right)_a^b"

[<Fact>]
let textWithExpression() =
    verifyParseResult @"\text{2+2}"

[<Theory>]
[<InlineData("{red}1123");
  InlineData("{red}{1}123");
  InlineData("{red} 1123");
  InlineData("{red} {1}123")>]
let color(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\color%s" text

[<Theory>]
[<InlineData(@"{red}1123");
  InlineData(@"{red}{1}123");
  InlineData(@"{red} 1123");
  InlineData(@"{red} {1}123")>]
let colorbox(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\colorbox%s" text

[<Theory>]
[<InlineData("2x123");
  InlineData("2{x}123");
  InlineData("{2}x123");
  InlineData("{2}{x}123");
  InlineData(" 2 x123");
  InlineData("2 {x}123");
  InlineData(" 2{x}123")>]
let frac(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\frac%s" text

[<Theory>]
[<InlineData("1123");
  InlineData("{1}123");
  InlineData(" 1123");
  InlineData(" {1}123")>]
let overline(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\overline%s" text

[<Theory>]
[<InlineData("1123");
  InlineData("{1}123");
  InlineData(" 1123");
  InlineData(" {1}123")>]
let sqrt(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\sqrt%s" text

[<Theory>]
[<InlineData(" [2]1123");
  InlineData(" [ 2]{1}123");
  InlineData("[2 ] 1123");
  InlineData("[ 2 ] {1}123")>]
let sqrtWithOptArgument(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\sqrt%s" text

[<Theory>]
[<InlineData("1123");
  InlineData("{1}123");
  InlineData(" 1123");
  InlineData(" {1}123")>]
let underline(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\underline%s" text

[<Theory>]
[<InlineData("x^y_z");
  InlineData("x^y_{z}");
  InlineData("x^{y}_z");
  InlineData("x^{y}_{z}");
  InlineData("x^y_ z");
  InlineData("x ^ {y} _ {z}")>]
let scripts(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(" 1123");
  InlineData(" {1}123")>]
let textArgumentParsing(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\text%s" text

[<Fact>]
let hat() : unit =
    verifyParseResult @"{\hat T}"

[<Fact>]
let integral() =
    verifyParseResult @"\int_a^b"
