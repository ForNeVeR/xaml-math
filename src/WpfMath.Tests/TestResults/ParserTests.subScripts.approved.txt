{
  "RootAtom": {
    "[AtomType]": "RowAtom",
    "Elements": [
      {
        "[AtomType]": "AccentedAtom",
        "AccentAtom": {
          "[AtomType]": "SymbolAtom",
          "IsDelimeter": false,
          "IsTextSymbol": false,
          "Name": "hat",
          "Source": null,
          "Type": "Accent"
        },
        "BaseAtom": {
          "[AtomType]": "CharAtom",
          "Character": "a",
          "IsTextSymbol": false,
          "Source": {
            "End": 6,
            "Length": 1,
            "Source": "\\hat{a}_{b\\sigma}",
            "SourceName": "User input",
            "Start": 5
          },
          "TextStyle": null,
          "Type": "Ordinary"
        },
        "Source": {
          "End": 4,
          "Length": 4,
          "Source": "\\hat{a}_{b\\sigma}",
          "SourceName": "User input",
          "Start": 0
        },
        "Type": "Ordinary"
      },
      {
        "[AtomType]": "ScriptsAtom",
        "BaseAtom": {
          "[AtomType]": "RowAtom",
          "Elements": [],
          "PreviousAtom": null,
          "Source": {
            "End": 17,
            "Length": 17,
            "Source": "\\hat{a}_{b\\sigma}",
            "SourceName": "User input",
            "Start": 0
          },
          "Type": "Ordinary"
        },
        "Source": {
          "End": 17,
          "Length": 10,
          "Source": "\\hat{a}_{b\\sigma}",
          "SourceName": "User input",
          "Start": 7
        },
        "SubscriptAtom": {
          "[AtomType]": "RowAtom",
          "Elements": [
            {
              "[AtomType]": "CharAtom",
              "Character": "b",
              "IsTextSymbol": false,
              "Source": {
                "End": 10,
                "Length": 1,
                "Source": "\\hat{a}_{b\\sigma}",
                "SourceName": "User input",
                "Start": 9
              },
              "TextStyle": null,
              "Type": "Ordinary"
            },
            {
              "[AtomType]": "SymbolAtom",
              "IsDelimeter": false,
              "IsTextSymbol": false,
              "Name": "sigma",
              "Source": {
                "End": 16,
                "Length": 5,
                "Source": "\\hat{a}_{b\\sigma}",
                "SourceName": "User input",
                "Start": 11
              },
              "Type": "Ordinary"
            }
          ],
          "PreviousAtom": null,
          "Source": {
            "End": 17,
            "Length": 16,
            "Source": "\\hat{a}_{b\\sigma}",
            "SourceName": "User input",
            "Start": 1
          },
          "Type": "Ordinary"
        },
        "SuperscriptAtom": null,
        "Type": "Ordinary"
      }
    ],
    "PreviousAtom": null,
    "Source": {
      "End": 17,
      "Length": 17,
      "Source": "\\hat{a}_{b\\sigma}",
      "SourceName": "User input",
      "Start": 0
    },
    "Type": "Ordinary"
  },
  "Source": {
    "End": 17,
    "Length": 17,
    "Source": "\\hat{a}_{b\\sigma}",
    "SourceName": "User input",
    "Start": 0
  },
  "TextStyle": null
}